src/libcore/mem.rs

```diff
@@ -415,6 +415,66 @@ pub fn replace<T>(dest: &mut T, mut src: T) -> T {
     src
 }
 
+/// A guarding type which will abort upon drop.
+///
+/// This is used for catching unwinding and transforming it into abort.
+struct ExitGuard;
+
+impl Drop for ExitGuard {
+    fn drop(&mut self) {
+        // To avoid unwinding, we abort the program, which ensures that the destructor of the
+        // invalidated value isn't runned.
+        unsafe { intrinsics::abort(); }
+    }
+}
+
+/// Temporarily takes ownership of a value at a mutable location, and replace it with a new value
+/// based on the old one.
+///
+/// We move out of reference temporarily, to apply a closure, returning a new value, which is then
+/// placed at the original value's location.
+///
+/// # An important note
+///
+/// The behavior on panic (or to be more precise, unwinding) is unspecified (but not undefined).
+///
+/// # Example
+///
+/// ```
+/// use std::mem;
+///
+/// // Dump some stuff into a box.
+/// let mut bx: Box<i32> = Box::new(200);
+///
+/// // Temporarily steal ownership.
+/// mem::replace(&mut bx, |mut owned| {
+///     owner = 5;
+///
+///     // The returned value is placed back in `&mut bx`.
+///     Box::new(owner)
+/// });
+/// ```
+#[inline]
+#[unstable(feature = "replace_with", issue = "...")]
+pub fn replace_with<T, F>(val: &mut T, replace: F)
+    where F: FnOnce(T) -> T {
+    // Guard against unwinding. Note that this is critical to safety, to avoid the value behind the
+    // reference `val` is not dropped twice during unwinding.
```

Comment:
Instead of "to avoid the value...is not dropped", maybe " to avoid the value...being dropped" or just "to avoid dropping the value..."